From: "Eric W. Biederman" <ebiederm@xmission.com>
Date: Tue, 7 Oct 2014 16:22:52 -0700
Subject: [PATCH] mnt: Move the clear of MNT_LOCKED from copy_tree to it's
 callers.
Git-commit: 8486a7882b5ba906992fd78bbfcefaae7fe285cc
Patch-mainline: v3.19
References: bnc#928547, CVE-2014-9717

Clear MNT_LOCKED in the callers of copy_tree except copy_mnt_ns, and
collect_mounts.  In copy_mnt_ns it is necessary to create an exact
copy of a mount tree, so not clearing MNT_LOCKED is important.
Similarly collect_mounts is used to take a snapshot of the mount tree
for audit logging purposes and auditing using a faithful copy of the
tree is important.

This becomes particularly significant when we start setting MNT_LOCKED
on rootfs to prevent it from being unmounted.

Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
Acked-by: NeilBrown <neilb@suse.com>

---
 fs/namespace.c |    1 -
 fs/pnode.c     |    1 +
 2 files changed, 1 insertion(+), 1 deletion(-)

--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1431,7 +1431,6 @@ struct mount *copy_tree(struct mount *mn
 	if (IS_ERR(q))
 		return q;
 
-	q->mnt.mnt_flags &= ~MNT_LOCKED;
 	q->mnt_mountpoint = mnt->mnt_mountpoint;
 
 	p = mnt;
--- a/fs/pnode.c
+++ b/fs/pnode.c
@@ -249,6 +249,7 @@ int propagate_mnt(struct mount *dest_mnt
 			list_splice(tree_list, tmp_list.prev);
 			goto out;
 		}
+		child->mnt.mnt_flags &= ~MNT_LOCKED;
 
 		if (is_subdir(dest_mp->m_dentry, m->mnt.mnt_root)) {
 			mnt_set_mountpoint(m, dest_mp, child);
